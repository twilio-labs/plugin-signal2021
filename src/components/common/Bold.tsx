import { Text, TextProps } from 'ink';
import React, { PropsWithChildren } from 'react';

export type BoldProps = TextProps;
export function Bold({ children, ...props }: PropsWithChildren<BoldProps>) {
  return (
    <Text bold color="redBright" {...props}>
      {children}
    </Text>
  );
}
