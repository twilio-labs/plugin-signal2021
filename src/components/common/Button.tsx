import { useFocus, useInput } from 'ink';
import React from 'react';
import { Merge } from 'type-fest';
import { Selectable, SelectableProps } from './Selectable';

export type ButtonProps = Merge<
  SelectableProps,
  {
    focusable?: boolean;
    autoFocus?: boolean;
    onAction?: () => void;
  }
>;
export function Button({
  children,
  focusable = true,
  autoFocus = true,
  onAction,
  ...props
}: ButtonProps) {
  const { isFocused } = useFocus({ isActive: focusable, autoFocus });
  useInput(
    (_, key) => {
      if (key.return && isFocused) {
        if (typeof onAction === 'function') {
          onAction();
        }
      }
    },
    {
      isActive: isFocused,
    }
  );
  return (
    <Selectable active={isFocused} {...props}>
      {children}
    </Selectable>
  );
}
