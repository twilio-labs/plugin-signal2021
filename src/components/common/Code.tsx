import { Text } from 'ink';
import React, { PropsWithChildren } from 'react';

export type CodeProps = PropsWithChildren<{}>;
export function Code({ children }: CodeProps) {
  return <Text color="redBright">{children}</Text>;
}
