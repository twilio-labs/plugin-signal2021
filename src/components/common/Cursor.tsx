import { Text } from 'ink';
import React from 'react';
import { Color } from '../../types/text';

export type CursorProps = {
  color?: Color;
  active?: boolean;
};
export function Cursor({ color = 'redBright', active = true }: CursorProps) {
  // using an empty space with background coloring because it renders more consistently than special characters
  return <Text backgroundColor={active ? color : undefined}> </Text>;
}
