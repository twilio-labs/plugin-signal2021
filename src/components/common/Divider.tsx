import { Box, Text } from 'ink';
import React, { useEffect, useState } from 'react';
import { useResponsiveMeasureElement } from '../../hooks/useResize';

function computeString(text: string, width: number, symbol: string) {
  if (symbol.length > 1) {
    throw new Error('Symbol can only be one character');
  }

  const paddingTotal = width - text.length - 2;
  const paddingSide = Math.floor(paddingTotal / 2);
  return `${symbol.repeat(paddingSide)} ${text} ${symbol.repeat(paddingSide)}`;
}

export type DividerProps = {
  title: string;
  symbol?: string;
};
export const Divider = ({ title, symbol = '-' }: DividerProps) => {
  const [text, setText] = useState(title);
  const { width, ref } = useResponsiveMeasureElement();

  useEffect(() => {
    if (!isNaN(width) && width > 0) {
      setText(computeString(title, width, symbol));
    }
  }, [title, width]);

  return (
    <Box ref={ref} width="100%">
      <Text>{text}</Text>
    </Box>
  );
};
