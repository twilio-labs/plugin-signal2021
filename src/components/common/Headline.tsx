import CFonts from 'cfonts';
import { Text } from 'ink';
import React, { useMemo } from 'react';
import { SmallHeadline } from './SmallHeadline';

function renderBigHeadline(
  text: string,
  size?: { width: number; height: number }
) {
  return CFonts.render(
    text,
    {
      maxLength: size.width,
    },
    false,
    2,
    {
      width: size.width,
      height: size.height,
    }
  ).string;
}

export type HeadlinePops = {
  text: string;
  shouldRenderLargeHeadline?: boolean;
  windowWidth?: number;
  windowHeight?: number;
};

export function Headline({
  text,
  shouldRenderLargeHeadline = true,
  windowWidth,
  windowHeight,
}: HeadlinePops) {
  const size =
    windowWidth && windowHeight
      ? { width: windowWidth, height: windowHeight }
      : undefined;
  const headline = useMemo(() => renderBigHeadline(text, size), [
    text,
    windowWidth,
    windowHeight,
  ]);

  if (!shouldRenderLargeHeadline) {
    return <SmallHeadline text={text} />;
  }
  return <Text>{headline}</Text>;
}
