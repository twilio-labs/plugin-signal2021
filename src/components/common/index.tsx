export * from './Bold';
export * from './Button';
export * from './Code';
export * from './Divider';
export * from './Info';
export * from './Inverse';
export * from './Key';
export * from './ListItem';
export * from './ListSelector';
export * from './LoadingIndicator';
export * from './OrderedList';
export * from './ProgrammingLanguageTag';
export * from './Selectable';
export * from './UnorderedList';
export * from './VerticalSpacer';
