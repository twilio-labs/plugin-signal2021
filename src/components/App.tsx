import { ApolloProvider } from '@apollo/client';
import { Box, render as renderApp, Text } from 'ink';
import useStdoutDimensions from 'ink-use-stdout-dimensions';
import React from 'react';
import { ModeProvider } from '../context/mode';
import { User, UserProvider } from '../context/user';
import { SHOW_APP } from '../utils/breakpoints';
import { client } from '../utils/graphqlClient';
import { Bold } from './common/Bold';
import { ITermWarning } from './dialogs/ITermWarning';
import { CheckUpdateAvailable } from './dialogs/UpdateNotice';
import { RenderIfWindowSize } from './media-breakpoints/RenderIfWindowSize';
import { ModeSwitcher } from './ModeSwitcher';
import { MainPane } from './panes/MainPane';
import { Sidebar } from './sidebar/Sidebar';

type AppProps = {
  user: Partial<User>;
};

export function App({ user }: AppProps) {
  const [columns, rows] = useStdoutDimensions();

  const appFallbackMessage = (
    <Box
      flexDirection="column"
      height={rows}
      width={columns}
      justifyContent="center"
      alignItems="center"
    >
      <Box width="50%" minWidth={30} flexDirection="column">
        <Bold>Thanks for entering SIGNAL Developer Mode.</Bold>
        <Text>
          Developer Mode will automatically adapt to your terminal window size
          but requires a certain minimum height to work well. Please either
          increase the size of your window or reduce the text size of your
          terminal.
        </Text>
      </Box>
    </Box>
  );

  return (
    <ApolloProvider client={client()}>
      <UserProvider defaultValue={user}>
        <ModeProvider>
          <ModeSwitcher />
          <RenderIfWindowSize {...SHOW_APP} fallback={appFallbackMessage}>
            <Box
              flexDirection="column"
              height={rows}
              width={columns}
              padding={1}
            >
              <CheckUpdateAvailable />
              <ITermWarning />
              <Box flexDirection="row" width="100%" flexGrow={1}>
                <Sidebar />
                <MainPane />
              </Box>
            </Box>
          </RenderIfWindowSize>
        </ModeProvider>
      </UserProvider>
    </ApolloProvider>
  );
}

export const render = async (
  { name, accountSid, twilioUsername, twilioPassword }: Partial<User>,
  debug = false
) => {
  process.title = 'SIGNAL Developer Mode';
  if (!debug) {
    const enterAltScreenCommand = '\x1b[?1049h';
    process.stdout.write(enterAltScreenCommand);
  }
  const { waitUntilExit } = renderApp(
    <App user={{ name, accountSid, twilioUsername, twilioPassword }} />,
    {
      debug: debug,
    }
  );

  await waitUntilExit();
  process.exit(0);
};
