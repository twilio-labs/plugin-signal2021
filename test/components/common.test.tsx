import { render } from 'ink-testing-library';
import React from 'react';
import { Bold, ListSelector } from '../../src/components/common';

describe('<Bold />', () => {
  it('should match snapshot', () => {
    const actual = render(<Bold>Hello</Bold>);
    expect(actual.lastFrame()).toMatchSnapshot();
  });
});

describe('<ListSelector />', () => {
  it('should match snapshot', () => {
    const actual = render(<ListSelector active />);
    expect(actual.lastFrame()).toMatchSnapshot();
  });
});
