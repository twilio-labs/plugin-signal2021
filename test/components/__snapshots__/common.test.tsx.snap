// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Bold /> should match snapshot 1`] = `"[1m[91mHello[39m[22m"`;

exports[`<ListSelector /> should match snapshot 1`] = `"[91m‚ùØ [39m"`;
